### DISCUSSION TITLE
api route inside gssp

### DISCUSSION THREAD

**Discussion by mrguamos** [vercel/next.js#41538] [Help]
In Nuxt3 they have something like this

> There are instances where you may need to directly call the API. Nuxt 3 provides a globally available $fetch method using [unjs/ohmyfetch](https://github.com/unjs/ohmyfetch) (in addition to fetch) with the same API as the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch).
> Using $fetch has a number of benefits, including:
> It will handle 'smartly' making direct API calls if it's running on the server, or making a client-side call to your API if it's running on the client. (It can also handle calling third-party APIs.)
> Plus, it comes with convenience features including automatically parsing responses and stringifying data.

and you can pass the cookie using 

`const { data } = await useFetch('/api/me', {
  headers: useRequestHeaders(['cookie'])
})`

i was just wondering if we can implement something like this in NextJS so we can call the api route inside GSSP and pass the cookie and use it inside the api route  (catch all) and add the desired headers and proxy it to an external api

**Comment by icyJoseph** [comment]
Hi,

This is a matter of architecture, that needs to be handled from your side. 

The thing is that reaching for an API route from GSSP, in this architecture, is rather inefficient and unnecessary, because it is an extra round-trip to get data, which you could already get from your GSSP function anyway.

Think of GSSP as a specialized API route, which rather than responding to the request, passes its result, as props, to the Page component associated with it.

Take look here, https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes:

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

So you need to design your integrations with 3rd parties so that they can be invoked by API routes, and data fetching functions.

For instance:

```ts
export const getCodeWarsUser = (user) =>
  fetch("https://www.codewars.com/api/v1/users/" + user)
    .then(res => res.json());
```

Now from wherever we want to call `getCodeWarsUser` we just need to calculate the user parameter. We can derive that from the incoming request in `GSSP`, and get the data directly from `GSSP` avoiding the unnecessary round trip to an API route. Thanks to Next.js' fetch polyfill (or Node 18+), the above can be called from browser, edge, or server.

```ts
export const getServerSideProps = async (ctx: GetServerSidePropsContext) => {
  console.log(ctx.req.headers);
  console.log(ctx.req.cookies);
  console.log(ctx.query);
  const user = ctx.query.user;

  if (!user) return { notFound: true };

  const props = await getCodeWarsUser(user);
  return { props };
};
```

I left a bunch of logs and the type signature, which can help you figure out what's available to you on GSSP. 

So, all you gotta do is refactor your API route so that the integration with 3rd parties is parametrised, and use that parametrised integration inside GSSP. 

Though, since this is JavaScript, you could still pass the request into your integration function, or call the API handler function directly with some dummy request/response. I wouldn't recommend this approach.

If you are worried about leaking to client, you can use https://next-code-elimination.vercel.app/ to make sure that code used only in GSSP is dropped from the client bundle.
